cluster/gce/windows/common.psm1

```diff
@@ -128,53 +128,132 @@ function Validate-SHA {
 # It will loop through the URLs list forever until it has a success. If
 # successful, it will write the file to OutFile. You can optionally provide a
 # Hash argument with an optional Algorithm, in which case it will attempt to
-# validate the downloaded file against the hash. SHA512 will be used if Algorithm
-# is not provided.
+# validate the downloaded file against the hash. SHA512 will be used if
+# -Algorithm is not provided.
+# This function is idempotent, if OutFile already exists and has the correct Hash
+# then the download will be skipped. If the Hash is incorrect, the file will be
+# overwritten.
 function MustDownload-File {
   param (
-    [parameter(Mandatory=$false)] [string]$Hash,
-    [parameter(Mandatory=$false)] [string]$Algorithm = 'SHA512',
-    [parameter(Mandatory=$true)] [string]$OutFile,
-    [parameter(Mandatory=$true)] [System.Collections.Generic.List[String]]$URLs,
-    [parameter(Mandatory=$false)] [System.Collections.IDictionary]$Headers = @{}
+    [parameter(Mandatory = $false)] [string]$Hash,
+    [parameter(Mandatory = $false)] [string]$Algorithm = 'SHA512',
+    [parameter(Mandatory = $true)] [string]$OutFile,
+    [parameter(Mandatory = $true)] [System.Collections.Generic.List[String]]$URLs,
+    [parameter(Mandatory = $false)] [System.Collections.IDictionary]$Headers = @{},
+    [parameter(Mandatory = $false)] [int]$Attempts = 0
   )
 
-  While($true) {
-    ForEach($url in $URLs) {
-      # If the URL is for GCS and the node has dev storage scope, add the
-      # service account token to the request headers.
-      if (($url -match "^https://storage`.googleapis`.com.*") -and $(Check-StorageScope)) {
-        $Headers["Authorization"] = "Bearer $(Get-Credentials)"
-      }
+  # If the file is already downloaded and matches the expected hash, skip the download.
+  if ((Test-Path -Path $OutFile) -And -Not [string]::IsNullOrEmpty($Hash)) {
+    try {
+      Validate-SHA -Hash $Hash -Path $OutFile -Algorithm $Algorithm
+      Log-Output "Skip download of ${OutFile}, it already exists with expected hash."
+      return
+    }
+    catch {
+      # The hash does not match the file on disk.
+      # Proceed with the download and overwrite the file.
+      Log-Output "${OutFile} exists but had wrong hash. Redownloading."
+    }
+  }
 
-      # Attempt to download the file
-      Try {
-        # TODO(mtaufen): When we finally get a Windows version that has Powershell 6
-        # installed we can set `-MaximumRetryCount 6 -RetryIntervalSec 10` to make this even more robust.
-        $result = Invoke-WebRequest $url -Headers $Headers -OutFile $OutFile -TimeoutSec 300
-      } Catch {
+  $currentAttempt = 0
+  while ($true) {
+    foreach ($url in $URLs) {
+      if (($Attempts -ne 0) -And ($currentAttempt -Gt 5)) {
```

Comment:
Use lowercase for -and and -gt? This is what the style guide suggests, and would be consistent with the rest of this code I think.